File: src/laserion/polarization.py
import numpy as np

# %% ################## ####
#### Polarization class ####
#### ################## ####


class Polarization:
    """
    General polarization state using a simple Jones-style description:

        p = p1 * e1 + p2 * exp(i * delta) * e2

    where e1, e2 are orthonormal basis vectors (in 3D),
    p1, p2 are real amplitudes, and delta is the phase lag of e2 relative to e1.
    """

    def __init__(self, e1, e2, p1=1.0, p2=0.0, delta=0.0):
        e1 = np.asarray(e1, dtype=float)
        e2 = np.asarray(e2, dtype=float)

        # normalize & (assume) orthogonal
        self.e1 = e1 / np.linalg.norm(e1)
        self.e2 = e2 / np.linalg.norm(e2)

        self.p1 = float(p1)
        self.p2 = float(p2)
        self.delta = float(delta)

    def vector(self, phase: float) -> np.ndarray:
        """
        Return the *real* polarization vector at given carrier phase.

        E ∝ Re[ (p1 e1 + p2 e^{i δ} e2) e^{i phase} ].
        """
        c = np.exp(1j * phase)  # carrier
        v_complex = self.p1 * self.e1 * c + self.p2 * np.exp(1j * self.delta) * self.e2 * c
        return np.real(v_complex)

    def is_pure_linear(self, tol=1e-12) -> bool:
        """
        Rough test: p2 ≈ 0 => effectively linear along e1.
        Useful if you want to restrict some approximations to linear pol.
        """
        return abs(self.p2) < tol

    @classmethod
    def from_lab_direction(cls, k_vec, pol_hint, delta=0.0):
        """
        Build a Polarization object for a beam with propagation k_vec (lab),
        using a lab-frame 'hint' pol_hint (e.g. (1,0,0) for 'x-like').

        The resulting polarization is strictly transverse to k̂ and uses a
        lab-consistent basis (e1,e2) derived from k̂ and the lab z-axis.

        Parameters
        ----------
        k_vec : array_like, shape (3,)
            Propagation direction in lab frame.
        pol_hint : array_like, shape (3,)
            Desired polarization direction in lab frame (will be projected
            onto the transverse plane).
        delta : float
            Phase lag between e1 and e2 (for elliptical / circular states).
            For linear polarization, use delta = 0.
        """
        k_vec = np.asarray(k_vec, dtype=float)
        k_hat = k_vec / np.linalg.norm(k_vec)

        # --- build beam-frame transverse basis (e1,e2) ---
        z_hat = np.array([0.0, 0.0, 1.0])
        if abs(np.dot(z_hat, k_hat)) < 0.999999:
            e1 = z_hat - np.dot(z_hat, k_hat) * k_hat
            e1 /= np.linalg.norm(e1)
        else:
            e1 = np.array([1.0, 0.0, 0.0])  # fallback if k ∥ z

        e2 = np.cross(k_hat, e1)
        e2 /= np.linalg.norm(e2)

        # --- project pol_hint onto transverse plane ---
        ph = np.asarray(pol_hint, dtype=float)
        if np.linalg.norm(ph) == 0:
            raise ValueError("pol_hint cannot be zero.")

        ph_perp = ph - np.dot(ph, k_hat) * k_hat
        if np.linalg.norm(ph_perp) < 1e-12:
            # Nearly parallel to k: fall back to e1
            ph_perp = e1.copy()
        ph_perp /= np.linalg.norm(ph_perp)

        # components in (e1,e2)
        p1 = float(np.dot(ph_perp, e1))
        p2 = float(np.dot(ph_perp, e2))
        norm_p = np.hypot(p1, p2)
        if norm_p > 0:
            p1 /= norm_p
            p2 /= norm_p

        return cls(e1=e1, e2=e2, p1=p1, p2=p2, delta=delta)


class LinearPolarization(Polarization):
    """
    Linear polarization in the transverse plane of a beam.

    For a beam with propagation direction k_vec, we build a transverse basis
    (e1_base, e2_base) tied to the lab frame (using z as a reference if possible).
    The polarization is then:

        E ∝ cos(angle) * e1_base + sin(angle) * e2_base,

    with 'angle' measured in that transverse plane. For k_vec || z, e1_base = x,
    e2_base = y, so angle=0 → x, angle=90° → y, as in the standard case.
    """

    def __init__(self, k_vec, angle: float = 0.0):
        k_vec = np.asarray(k_vec, dtype=float)
        k_hat = k_vec / np.linalg.norm(k_vec)

        # Build transverse basis (e1_base, e2_base) tied to lab frame
        z_hat = np.array([0.0, 0.0, 1.0])
        if abs(np.dot(z_hat, k_hat)) < 0.999999:
            e1_base = z_hat - np.dot(z_hat, k_hat) * k_hat
            e1_base /= np.linalg.norm(e1_base)
        else:
            # If k is (anti-)parallel to z, fall back to x as reference
            e1_base = np.array([1.0, 0.0, 0.0])

        e2_base = np.cross(k_hat, e1_base)
        e2_base /= np.linalg.norm(e2_base)

        # Angle in radians
        theta = angle * np.pi / 180.0

        # Polarization direction in the transverse plane
        # e_pol = cos θ e1_base + sin θ e2_base
        p1 = np.cos(theta)
        p2 = np.sin(theta)

        super().__init__(e1=e1_base, e2=e2_base, p1=p1, p2=p2, delta=0.0)


class CircularPolarization(Polarization):
    """
    Circular polarization in the transverse plane of a beam.

    For a beam with propagation direction k_vec, we build a transverse basis
    (e1_base, e2_base) tied to the lab frame (using z as a reference if possible).

    We then optionally rotate this basis by 'angle' in the transverse plane:

        e1 = cos(angle) * e1_base + sin(angle) * e2_base
        e2 = -sin(angle) * e1_base + cos(angle) * e2_base

    and define circular polarization with equal amplitudes p1 = p2 = 1 and
    phase lag δ = ±π/2 between e1 and e2.
    """

    def __init__(self, k_vec, sense: str = "right", angle: float = 0.0):
        k_vec = np.asarray(k_vec, dtype=float)
        k_hat = k_vec / np.linalg.norm(k_vec)

        # Build transverse basis (e1_base, e2_base) tied to lab frame
        z_hat = np.array([0.0, 0.0, 1.0])
        if abs(np.dot(z_hat, k_hat)) < 0.999999:
            e1_base = z_hat - np.dot(z_hat, k_hat) * k_hat
            e1_base /= np.linalg.norm(e1_base)
        else:
            e1_base = np.array([1.0, 0.0, 0.0])

        e2_base = np.cross(k_hat, e1_base)
        e2_base /= np.linalg.norm(e2_base)

        # Rotate the transverse basis by 'angle' in that plane
        theta = angle * np.pi / 180.0
        cos_t = np.cos(theta)
        sin_t = np.sin(theta)

        e1 = cos_t * e1_base + sin_t * e2_base
        e2 = -sin_t * e1_base + cos_t * e2_base

        # Choose phase lag for handedness
        sense_l = sense.lower()
        if sense_l in ("right", "cw", "clockwise"):
            delta = +np.pi / 2.0
        elif sense_l in ("left", "ccw", "counterclockwise", "counter-clockwise"):
            delta = -np.pi / 2.0
        else:
            raise ValueError(f"Unknown circular polarization sense: {sense!r}")

        # Circular: equal amplitudes along e1, e2, ±π/2 phase lag
        super().__init__(e1=e1, e2=e2, p1=1.0, p2=1.0, delta=delta)

File: csrc/src/polarization.c
/* csrc/src/polarization.c */
#include <math.h>
#include <string.h>
#include <ctype.h>
#include "polarization.h"

/* -------------------- small vector helpers -------------------- */

static double vdot(const double a[3], const double b[3])
{
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
}

static void vcross(const double a[3], const double b[3], double out[3])
{
    out[0] = a[1] * b[2] - a[2] * b[1];
    out[1] = a[2] * b[0] - a[0] * b[2];
    out[2] = a[0] * b[1] - a[1] * b[0];
}

static double vnorm(const double a[3])
{
    return sqrt(vdot(a, a));
}

static int vnormalize(const double in[3], double out[3])
{
    double n = vnorm(in);
    if (n <= 0.0)
        return 1;
    out[0] = in[0] / n;
    out[1] = in[1] / n;
    out[2] = in[2] / n;
    return 0;
}

static void vsub_scaled(double out[3], const double a[3], const double b[3], double s)
{
    /* out = a - s*b */
    out[0] = a[0] - s * b[0];
    out[1] = a[1] - s * b[1];
    out[2] = a[2] - s * b[2];
}

/* case-insensitive string compare for small tokens */
static int str_ieq(const char *a, const char *b)
{
    while (*a && *b)
    {
        if (tolower((unsigned char)*a) != tolower((unsigned char)*b))
            return 0;
        ++a;
        ++b;
    }
    return (*a == '\0' && *b == '\0');
}

/* Build lab-tied transverse basis (e1_base, e2_base) for a given k_hat. */
static int build_transverse_basis(const double k_hat[3],
                                  double e1_base[3],
                                  double e2_base[3])
{
    const double z_hat[3] = {0.0, 0.0, 1.0};

    if (fabs(vdot(z_hat, k_hat)) < 0.999999)
    {
        /* e1_base = z_hat - (z_hat·k_hat) k_hat */
        vsub_scaled(e1_base, z_hat, k_hat, vdot(z_hat, k_hat));
        if (vnormalize(e1_base, e1_base) != 0)
            return 3;
    }
    else
    {
        /* fallback if k || z */
        const double x_hat[3] = {1.0, 0.0, 0.0};
        e1_base[0] = x_hat[0];
        e1_base[1] = x_hat[1];
        e1_base[2] = x_hat[2];
    }

    vcross(k_hat, e1_base, e2_base);
    if (vnormalize(e2_base, e2_base) != 0)
        return 4;

    return 0;
}

/* Rotate (e1_base,e2_base) by angle_deg in the transverse plane. */
static void rotate_basis(const double e1_base[3],
                         const double e2_base[3],
                         double angle_deg,
                         double e1[3],
                         double e2[3])
{
    const double pi = 3.14159265358979323846;
    double theta = angle_deg * (pi / 180.0);
    double c = cos(theta);
    double s = sin(theta);

    /* e1 =  c e1_base + s e2_base
       e2 = -s e1_base + c e2_base */
    e1[0] = c * e1_base[0] + s * e2_base[0];
    e1[1] = c * e1_base[1] + s * e2_base[1];
    e1[2] = c * e1_base[2] + s * e2_base[2];

    e2[0] = -s * e1_base[0] + c * e2_base[0];
    e2[1] = -s * e1_base[1] + c * e2_base[1];
    e2[2] = -s * e1_base[2] + c * e2_base[2];
}

/* sense -> delta (+pi/2 right-handed, -pi/2 left-handed) */
static int sense_to_delta(const char *sense, double *delta_out)
{
    const double pi = 3.14159265358979323846;
    if (!sense || !delta_out)
        return 1;

    if (str_ieq(sense, "right") || str_ieq(sense, "cw") || str_ieq(sense, "clockwise"))
    {
        *delta_out = +pi / 2.0;
        return 0;
    }
    if (str_ieq(sense, "left") || str_ieq(sense, "ccw") ||
        str_ieq(sense, "counterclockwise") || str_ieq(sense, "counter-clockwise"))
    {
        *delta_out = -pi / 2.0;
        return 0;
    }
    return 2;
}

/* -------------------- public API -------------------- */

int Polarization_init(Polarization *p,
                      const double e1[3],
                      const double e2[3],
                      double p1,
                      double p2,
                      double delta)
{
    if (!p)
        return 1;

    if (vnormalize(e1, p->e1) != 0)
        return 2;
    if (vnormalize(e2, p->e2) != 0)
        return 3;

    p->p1 = p1;
    p->p2 = p2;
    p->delta = delta;
    return 0;
}

void Polarization_vector(const Polarization *p,
                         double phase,
                         double out_vec[3])
{
    /* Re[ (p1 e1 + p2 e^{i delta} e2) e^{i phase} ]
     * = p1 e1 cos(phase) + p2 e2 cos(phase + delta)
     */
    double c1 = cos(phase);
    double c2 = cos(phase + p->delta);

    out_vec[0] = p->p1 * p->e1[0] * c1 + p->p2 * p->e2[0] * c2;
    out_vec[1] = p->p1 * p->e1[1] * c1 + p->p2 * p->e2[1] * c2;
    out_vec[2] = p->p1 * p->e1[2] * c1 + p->p2 * p->e2[2] * c2;
}

bool Polarization_is_pure_linear(const Polarization *p, double tol)
{
    return fabs(p->p2) < tol;
}

int Polarization_make_basis_angle(const double k_vec[3],
                                  double angle_deg,
                                  double e1_out[3],
                                  double e2_out[3])
{
    if (!k_vec || !e1_out || !e2_out)
        return 1;

    /* k_hat */
    double k_hat[3];
    if (vnormalize(k_vec, k_hat) != 0)
        return 2;

    double e1_base[3], e2_base[3];
    {
        int rc = build_transverse_basis(k_hat, e1_base, e2_base);
        if (rc != 0)
            return rc;
    }

    rotate_basis(e1_base, e2_base, angle_deg, e1_out, e2_out);
    return 0;
}

int Polarization_from_lab_direction(Polarization *p,
                                    const double k_vec[3],
                                    const double pol_hint[3],
                                    double delta)
{
    if (!p)
        return 1;

    /* k_hat */
    double k_hat[3];
    if (vnormalize(k_vec, k_hat) != 0)
        return 2;

    /* base transverse basis tied to lab */
    double e1_base[3], e2_base[3];
    {
        int rc = build_transverse_basis(k_hat, e1_base, e2_base);
        if (rc != 0)
            return rc;
    }

    /* project pol_hint onto plane perpendicular to k */
    if (vnorm(pol_hint) == 0.0)
        return 5;

    double ph_perp[3];
    vsub_scaled(ph_perp, pol_hint, k_hat, vdot(pol_hint, k_hat));

    if (vnorm(ph_perp) < 1e-12)
    {
        /* nearly parallel to k: fall back to e1_base */
        ph_perp[0] = e1_base[0];
        ph_perp[1] = e1_base[1];
        ph_perp[2] = e1_base[2];
    }
    else
    {
        if (vnormalize(ph_perp, ph_perp) != 0)
            return 6;
    }

    /* components in (e1_base, e2_base) */
    double p1 = vdot(ph_perp, e1_base);
    double p2 = vdot(ph_perp, e2_base);

    double norm_p = hypot(p1, p2);
    if (norm_p > 0.0)
    {
        p1 /= norm_p;
        p2 /= norm_p;
    }

    return Polarization_init(p, e1_base, e2_base, p1, p2, delta);
}

int LinearPolarization_init(Polarization *p,
                            const double k_vec[3],
                            double angle_deg)
{
    if (!p)
        return 1;

    double e1[3], e2[3];
    int rc = Polarization_make_basis_angle(k_vec, angle_deg, e1, e2);
    if (rc != 0)
        return rc;

    /* linear along e1 (after rotation) */
    return Polarization_init(p, e1, e2, 1.0, 0.0, 0.0);
}

int CircularPolarization_init(Polarization *p,
                              const double k_vec[3],
                              const char *sense,
                              double angle_deg)
{
    if (!p)
        return 1;

    double e1[3], e2[3];
    int rc = Polarization_make_basis_angle(k_vec, angle_deg, e1, e2);
    if (rc != 0)
        return rc;

    double delta;
    rc = sense_to_delta(sense, &delta);
    if (rc != 0)
        return 6;

    return Polarization_init(p, e1, e2, 1.0, 1.0, delta);
}
